output/cloud/expv2/metrics_client.go

```diff
@@ -73,33 +77,73 @@ func (mc *metricsClient) push(ctx context.Context, referenceID string, samples *
 	// we don't expect to share this client across different refID
 	// with a bit of effort we can find a way to just allocate once
 	url := mc.baseURL + referenceID
-	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(b))
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, io.NopCloser(bytes.NewReader(b)))
 	if err != nil {
 		return err
 	}
 
+	req.GetBody = func() (io.ReadCloser, error) {
+		return io.NopCloser(bytes.NewReader(b)), nil
+	}
+
 	req.Header.Set("User-Agent", mc.userAgent)
 	req.Header.Set("Content-Type", "application/x-protobuf")
 	req.Header.Set("Content-Encoding", "snappy")
 	req.Header.Set("K6-Metrics-Protocol-Version", "2.0")
 	req.Header.Set("Authorization", "Token "+mc.token)
 
-	resp, err := mc.httpClient.Do(req)
+	err = mc.do(req)
 	if err != nil {
 		return err
 	}
-	defer func() {
-		_ = resp.Body.Close()
-	}()
 
-	if err := cloudapi.CheckResponse(resp); err != nil {
-		return err
-	}
 	mc.logger.WithField("t", time.Since(start)).WithField("size", len(b)).
 		Debug("Pushed the collected metrics to the Cloud service")
 	return nil
 }
 
+func (mc *metricsClient) do(r *http.Request) error {
+	var (
+		err   error
+		retry = true
+	)
+
+	for attempts := 1; retry; attempts++ {
+		retry, err = mc.retryableDo(r, attempts)
+
+		if retry {
+			mc.waitFor(time.Duration(attempts) * retryIntervalFactor)
+			continue
+		}
+	}
+
+	return err
+}
+
+func (mc *metricsClient) retryableDo(req *http.Request, attempt int) (retry bool, err error) {
+	resp, err := mc.httpClient.Do(req)
+
+	defer func() {
+		if resp != nil {
+			_ = resp.Body.Close()
```

Comment:
You should also read it in full before you close it

Reply:
Done